Network: Simplify the bind logic

a1rwulf opened the pull request:
- increase readability of the code
 - use closesocket consistently
 - fix another potential sizeof issue

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [x] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

fritsch commented on xbmc/network/Socket.cpp:
```diff
@@ -171,6 +138,42 @@ bool CPosixUDPSocket::Bind(bool localOnly, int port, int range)
+  CAddress testaddr("::");
+
+  SOCKET testSocket = socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP);
+  if (testSocket == INVALID_SOCKET)
```
What's kodi's philosophy on resetting ```errno```?

a1rwulf replied:
I did a short spot check and couldn't find any code that resets `errno`.

fritsch commented on xbmc/network/Socket.cpp:
```diff
@@ -171,6 +138,42 @@ bool CPosixUDPSocket::Bind(bool localOnly, int port, int range)
+  for (; port <= port + range; port++)
+  {
+    testaddr.saddr.saddr6.sin6_port = htons(port);
+    if (bind(testSocket, reinterpret_cast<struct sockaddr*>(&testaddr.saddr), testaddr.size) >= 0)
```
bind should return 0 in success case, therefore the ``` >=``` is not needed. Also on failure this sets ```errno``` - just if we want to check that.

fritsch commented on xbmc/network/Socket.cpp:
```diff
@@ -171,6 +138,42 @@ bool CPosixUDPSocket::Bind(bool localOnly, int port, int range)
+
+  SOCKET testSocket = socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP);
+  if (testSocket == INVALID_SOCKET)
+    return false;
```
return ipv6? Or just use true, false alls the way.

fritsch commented on xbmc/network/Socket.cpp:
```diff
@@ -171,6 +138,42 @@ bool CPosixUDPSocket::Bind(bool localOnly, int port, int range)
+    testaddr.saddr.saddr6.sin6_port = htons(port);
+    if (bind(testSocket, reinterpret_cast<struct sockaddr*>(&testaddr.saddr), testaddr.size) >= 0)
+    {
+      ipv6 = true;
```
I see the ipv6 bool is just there, so that one can do another closesocket before returning. Scope Guard would be nice, but it's okay as is.

a1rwulf replied:
I put a commit on top that's using ScopeGuard, if we want to keep it I'll squash.
Otherwise I'll drop it, it's a but unfortunate that you can't feed it with the closesocket define :(

fritsch replied:
No, no. Keep it as is. The other commit can come on top, afterwards :-)

a1rwulf replied:
argh, there is still an error in it..

fritsch reviewed (approved):
My other comments are just cosmetics.

a1rwulf commented:
OK, hope this is the final version now.

a1rwulf commented:
jenkins build this please

yol commented on xbmc/network/Socket.cpp:
```diff
@@ -171,6 +139,56 @@ bool CPosixUDPSocket::Bind(bool localOnly, int port, int range)
+
+  if (static_cast<SOCKET>(testSocket) == INVALID_SOCKET)
+  {
+    CLog::Log(LOGDEBUG, "{} - Could not create IPv6 socket: %s", __FUNCTION__, strerror(errno));
```
```suggestion
    CLog::LogF(LOGDEBUG, "Could not create IPv6 socket: %s", strerror(errno));
```
(same for others)

yol commented on xbmc/network/Socket.cpp:
```diff
@@ -34,56 +35,23 @@ bool CPosixUDPSocket::Bind(bool localOnly, int port, int range)
-          closesocket(testSocket);
-          testSocket = INVALID_SOCKET;
         }
-        else
-        {
-          CLog::Log(LOGWARNING, "UDP: Could not create testSocket");
-          close(m_iSock);
-          m_iSock = INVALID_SOCKET;
-        }
-
-        if (testSocket != INVALID_SOCKET)
-          closesocket(testSocket);
       }
```
wrong indentation

a1rwulf committed:
Network: Simplify the bind logic

 - increase readability of the code
 - use closesocket consistently
 - fix another potential sizeof issue

a1rwulf commented:
jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins didn't start

Jenkins build this please
